messages2/src/decorators/attachment.rs

```diff
@@ -0,0 +1,211 @@
+    #[serde(rename = "mime-type")]
+    pub mime_type: Option<MimeType>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub lastmod_time: Option<DateTime<Utc>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub byte_count: Option<u64>,
+    pub data: AttachmentData,
+}
+
+impl Attachment {
+    pub fn new(data: AttachmentData) -> Self {
+        Self {
+            id: None,
+            description: None,
+            filename: None,
+            mime_type: None,
+            lastmod_time: None,
+            byte_count: None,
+            data,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct AttachmentData {
+    // There probably is a better type for this???
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub jws: Option<String>,
+    // Better type for this as well?
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub sha256: Option<String>,
+    #[serde(flatten)]
+    pub content: AttachmentType,
+}
+
+impl AttachmentData {
+    pub fn new(content: AttachmentType) -> Self {
+        Self {
+            jws: None,
+            sha256: None,
+            content,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "kebab-case")]
+pub enum AttachmentType {
+    Base64(String),
+    Json(Value),
+    Links(Vec<Url>),
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+#[allow(clippy::field_reassign_with_default)]
+pub mod tests {
+    use serde_json::json;
+
+    use super::*;
+    use crate::misc::test_utils;
+
+    pub fn make_minimal_attachment() -> Attachment {
+        let data = json!({
+            "field": "test_json_data"
+        });
+
+        let content = AttachmentType::Json(data);
+        let attach_data = AttachmentData::new(content);
+        Attachment::new(attach_data)
+    }
+
+    pub fn make_extended_attachment() -> Attachment {
+        let data = json!({
+            "field": "test_json_data"
+        });
+
+        let content = AttachmentType::Json(data);
+        let attach_data = AttachmentData::new(content);
+        let mut attachment = Attachment::new(attach_data);
+        let id = "test_id".to_owned();
+        let description = "test_description".to_owned();
+        let filename = "test_filename".to_owned();
+        let mime_type = MimeType::Json;
+        let lastmod_time = DateTime::<Utc>::default();
+        let byte_count = 64;
+
+        attachment.id = Some(id);
+        attachment.description = Some(description);
+        attachment.filename = Some(filename);
+        attachment.mime_type = Some(mime_type);
+        attachment.lastmod_time = Some(lastmod_time);
+        attachment.byte_count = Some(byte_count);
+
+        attachment
+    }
+
+    #[test]
+    fn test_base64_attach_type() {
+        let data = "test_base64_str".to_owned();
+
+        let json = json!({ "base64": data });
+        let attach_type = AttachmentType::Base64(data);
```

Comment:
Replaced `String` with `Vec<u8>`

Reply:
Well, the most safe, efficient and convenient solution would be to do the encoding upon setting it. That way messages would assure the encoding is correct without having to validate already encoded b64. Also convenient for user. 
We can leave that for later, maybe good first issue :-)

Reply:
We could, but that means locking all consumers into the same base64 encoding library. I think it's not a huge thing to allow users to pass arbitrary input as long as they realize that the other party won't understand it unless it is base64 encoded.

It would also be possible in some cases for users to already have base64 encoded data, so encoding it ourselves wouldn't be necessary.